Add CaDiCaL support to CMake

tautschnig opened the pull request:
Add CMake configuration to build using CaDiCaL, which was already
supported with Makefiles. Also bump the version of CaDiCaL to sr19 and
fix the use of the cnf_solvert interface.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Add CaDiCaL support to CMake

Add CMake configuration to build using CaDiCaL, which was already
supported with Makefiles. Also bump the version of CaDiCaL to rel-1.3.0
and fix the use of the cnf_solvert interface.

Run the ubuntu-18_04-package GitHub PR action with -Dsat_impl=cadical to
avoid future regressions.

Fixes: #5348

tautschnig commented:
FWIW: running the regression suite with CaDiCaL appears to be quite a bit slower, but an evaluation on more realistic benchmarks is yet to be done.

smowton commented on src/solvers/CMakeLists.txt:
```diff
@@ -98,6 +98,36 @@ elseif("${sat_impl}" STREQUAL "glucose")
+    download_project(PROJ cadical
+        URL https://github.com/arminbiere/cadical/archive/sr19.tar.gz
+        PATCH_COMMAND true
+        COMMAND CXX=${CXX_COMPILER} CXXFLAGS=-O3 ./configure --debug
```
Double-check: is the `--debug` asking `configure` for verbose output, or setting `-g` for the build, or could it even be overriding the `-O3` given in `CXXFLAGS` to get `-g -O0`?

tautschnig replied:
I did double-check when building locally: it just adds `-g`, the `-O3` from `CXXFLAGS` is passed on as desired.

smowton reviewed (approved):
Build script stuff looks fine to me, no opinion on the solvers stuff that I'm not that familiar with

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7931e35fe6f23fd32dd6227231d3df26c7089f28).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114606145

thk123 commented on src/solvers/sat/satcheck_cadical.cpp:
```diff
@@ -79,6 +79,21 @@ propt::resultt satcheck_cadicalt::do_prop_solve()
   }
   else
   {
+    // if assumptions contains false, we need this to be UNSAT
```
:pick: I don't understand this change or why the assume falses need special casing, but it should probably go in its own commit explaining in more detail why this is required

thk123 replied:
Preferably with a test that only passes with this

tautschnig replied:
Just for the record: this is done in keeping with other solvers, and thus about the way in which we use the solver interface. I agree that a test would be useful.

kroening replied:
BTW, the reason why 'assume false' needs special casing is that SAT solvers don't have a 'false' literal.

thk123 reviewed (commented):
I don't have any strong opinions on this, but it should probably build `cadical` on CI and run the tests?

peterschrammel commented on src/solvers/sat/satcheck_cadical.cpp:
```diff
@@ -120,13 +135,21 @@ satcheck_cadicalt::~satcheck_cadicalt()
+  {
+    if(a.is_true())
+    {
+      assumptions.clear();
```
:no_entry_sign: I don't understand this. Assumptions are connected conjunctively.

tautschnig replied:
That was me copy&pasting from the `satcheck_minisat2.cpp` ... probably worth reviewing.

peterschrammel replied:
See https://github.com/diffblue/cbmc/pull/4810

martin-cs reviewed (approved):
Seems good to me.

peterschrammel reviewed (approved):
I'd suggest to merge also https://github.com/diffblue/cbmc/pull/5305 so that we can use the other solvers too without having to implement hardness_collectort everywhere.

tautschnig committed:
BV refinement must not touch the formula before requesting the model

A SAT instance that is satisfiable need not remain so when adding
clauses. Solvers strictly asserting this (such as Cadical) would refuse
getting the model of a satisfiable instance after having added further
clauses since the last solve() call.

Thus, request the values of all approximations first, and only then add
further constraints. Doing so makes cbmc-library/Float-div1-refine pass
again, even when using Cadical.

hannes-steffenhagen-diffblue commented on src/solvers/CMakeLists.txt:
```diff
@@ -98,6 +98,36 @@ elseif("${sat_impl}" STREQUAL "glucose")
+
+    set_target_properties(
+        cadical
+        PROPERTIES IMPORTED_LOCATION ${cadical_SOURCE_DIR}/build/libcadical.a
```
This will only work on Linux, is that OK? If cadical also works on Windows/OSX it'd be nice if just used something like `find_library` here to get a more portable library path.

tautschnig replied:
Cadical's makefile hard-codes this, so I don't think it'll work out of the box on other platforms.

hannes-steffenhagen-diffblue commented on src/solvers/sat/satcheck_cadical.cpp:
```diff
@@ -23,7 +23,7 @@ tvt satcheck_cadicalt::l_get(literalt a) const
   tvt result;
 
-  if(a.var_no() > static_cast<unsigned>(solver->max()))
+  if(a.var_no() > static_cast<unsigned>(solver->vars()))
```
Preferably use `narrow` from `util/narrow.h` here

tautschnig replied:
Will do.

hannes-steffenhagen-diffblue commented on src/solvers/sat/satcheck_cadical.cpp:
```diff
@@ -72,31 +89,37 @@ propt::resultt satcheck_cadicalt::do_prop_solve()
   }
 
+  for(const auto &a : assumptions)
+    solver->assume(a.dimacs());
```
Why do we need to do this now? This wasn't here previously?

tautschnig replied:
Support for assumptions simply was missing before.

hannes-steffenhagen-diffblue reviewed (approved):
@tautschnig I'm noticing that the second commit has quite a few related changes at once. Is there any chance these could be split up into indivial commits that still make sense on their own? No real issue with the changes themselves.

tautschnig commented:
> @tautschnig I'm noticing that the second commit has quite a few related changes at once. Is there any chance these could be split up into indivial commits that still make sense on their own? No real issue with the changes themselves.

I'm not entirely convinced this adds value: the previous implementation was pretty broken, and not really usable. All these changes are necessary to make things work.
